lte/gateway/c/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp

```diff
@@ -0,0 +1,261 @@
+        break;
+        // many more to be coded
+      default:
+        *ul_pdu_ambr = (256 * smf_context->ul_session_ambr);
+        break;
+    }
+  }
+}
+
+/*
+ * As per Baicel capture and message flow, the function to be called
+ * before sending gRPC message to SMF with avilable information in
+ * smf_context or default.
+ * No NAS message is involved and direct itti_message is sent to NGAP.
+ * The transfer message is bstring and heavily nested structure is
+ * converted to bstring before sending to NGAP
+ */
+int pdu_session_resource_setup_request(
+    ue_m5gmm_context_s* ue_context, amf_ue_ngap_id_t amf_ue_ngap_id) {
+  if (!ue_context) {
+    // TODO
+  }
+  int rc = RETURNerror;
+  smf_context_t* smf_context;
+  amf_context_t* amf_context;
+  amf_context = &ue_context->amf_context;
+  smf_context = &ue_context->amf_context.smf_context;
+  pdu_session_resource_setup_request_transfer_t
+      amf_pdu_ses_setup_transfer_req;  // pdu_res_set_change
+  itti_ngap_pdusession_resource_setup_req_t* ngap_pdu_ses_setup_req = nullptr;
+  MessageDef* message_p                                             = nullptr;
+  uint64_t dl_pdu_ambr;
+  uint64_t ul_pdu_ambr;
+  message_p =
+      itti_alloc_new_message(TASK_AMF_APP, NGAP_PDUSESSION_RESOURCE_SETUP_REQ);
+  ngap_pdu_ses_setup_req =
+      &message_p->ittiMsg.ngap_pdusession_resource_setup_req;
+  memset(
+      ngap_pdu_ses_setup_req, 0,
+      sizeof(itti_ngap_pdusession_resource_setup_req_t));
+
+  // start filling message in DL to NGAP
+  ngap_pdu_ses_setup_req->gnb_ue_ngap_id = ue_context->gnb_ue_ngap_id;
+  ngap_pdu_ses_setup_req->amf_ue_ngap_id = amf_ue_ngap_id;
+
+  /*
+   * by this time amf and smf context available but ambr for pdu not available
+   * considering default or max bit rate.
+   * leveraged ambr calculation from qos_params_to_eps_qos and 24-501 spec used
+   */
+  ambr_calculation_pdu_session(smf_context, &dl_pdu_ambr, &ul_pdu_ambr);
+  ngap_pdu_ses_setup_req->ue_aggregate_maximum_bit_rate.dl = dl_pdu_ambr;
+  ngap_pdu_ses_setup_req->ue_aggregate_maximum_bit_rate.ul = ul_pdu_ambr;
+
+  // Hardcored number of pdu sessions as 1
+  ngap_pdu_ses_setup_req->pduSessionResource_setup_list.no_of_items = 1;
+  ngap_pdu_ses_setup_req->pduSessionResource_setup_list.item[0].Pdu_Session_ID =
+      (Ngap_PDUSessionID_t)
+          smf_context->smf_proc_data.pdu_session_identity.pdu_session_id;
+
+  /* preparing for  bstring PDU_Session_Resource_Setup_Transfer
+   * amf_pdu_ses_setup_transfer_req is the structure to be filed
+   * and make a bundle and converted to bstring
+   */
+  amf_pdu_ses_setup_transfer_req.pdu_aggregate_max_bit_rate.dl = dl_pdu_ambr;
+  amf_pdu_ses_setup_transfer_req.pdu_aggregate_max_bit_rate.ul = ul_pdu_ambr;
+
+  // UPF tied 4 octet and respective ip address are from
+  // SMF date that written to context
+  memcpy(
+      &amf_pdu_ses_setup_transfer_req.up_transport_layer_info.gtp_tnl.gtp_tied,
+      smf_context->gtp_tunnel_id.upf_gtp_teid, 4);
```

Comment:
check with @sanjay-ACL, this field will be coming as uint32_t from SMF. Make sure that it was converted to network byte order.

Reply:
we will update and raise a separate PR.